src/index.ts

```diff
@@ -29,9 +26,13 @@ export interface StoreApi<T> {
 }
 
 const reducer = <T>(state: any, newState: T) => newState
+const useIsomorphicLayoutEffect =
+  typeof window !== 'undefined' ? useLayoutEffect : useEffect
 
 export default function create<TState extends State>(
-  createState: (set: SetState<State>, get: GetState<State>, api: any) => TState
+  createState: keyof TState extends never
```

Comment:
Just to understand, is it to detect if we're in JS vs TS land?

I just tested it and it does fix the type inference issue in TS. We still have to pass the state generic explicitely to `create` to not get implicit anys but I think that's a good DX behaviour.

Reply:
> Just to understand, is it to detect if we're in JS vs TS land?

Essentially yeah. It checks to see if the state generic was passed in or not.

Reply:
Hi,
The fix works fine with a very basic sample like this one :
![Screen Shot 2019-07-03 at 17 55 54](https://user-images.githubusercontent.com/1252966/60606535-d2a53b80-9dbb-11e9-98fb-336b825f3519.png)

But I'm afraid that th fix is not enough because it is very easy to get a non working case. Imagine you have several stores, and imagine you want to do something generic on them. You will create a function that performs the "generic" things and then call that function for each store you have.

This will lead you to create a generic function that encapsulates the zustand `create` function invokation. But as far as I can see, it doesn't work.

Just take a look at this really trivial sample in which we try to do something generic during store intialization : 
![Screen Shot 2019-07-03 at 17 56 45](https://user-images.githubusercontent.com/1252966/60606601-f10b3700-9dbb-11e9-8d68-2b607d866bf6.png)

I hope you will agree with me that this sample code is trivial (10 lines of effective code), and it highlights the problem...

In other words, as soon as you want to add even a thin layer over zustand for your own purpose in Typescript world, you meet that problem.

So on my side, I still have the same issue as the one I raised in issue #32.

Reply:
I have put all this (trivial) code on codesandbox : https://codesandbox.io/s/j2nys
(`basic.ts` and `initStore.ts` files)

Reply:
You're right. I've tested without passing the generic `TState` from `initStore` and it works.

```ts
interface State {
  foo: string;
}

export function initStore(initialState: State) {
  return create<State>((set: SetState<State>, get: GetState<State>) => {
    return {
      foo: "bar"
    };
  });
}
```

But Typescript complain as soon as we pass a generic down the line:

```ts
export function initStore<S>(initialState: S) {
  return create<S>((set: SetState<S>, get: GetState<S>) => {
    return {};
  });
}
```

It looks like the conditional type isn't kicking in correctly when a generic is  used.